---
title: "Getting My Colleagues Hooked on R"
author: "Florian Priv√©"
date: "August 5, 2016"
output:
  ioslides_presentation:
    css: styles.css
    self_contained: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(autodep = TRUE, fig.align="center",
                      fig.retina = NULL)
```


## What is R? 

<iframe width="640" height="360" src="https://www.youtube.com/embed/TR2bHSJ_eck" frameborder="0" allowfullscreen></iframe>


## 7 facts about the growth of R:

```{r, echo = FALSE}
thepage <- readLines('https://cran.r-project.org/web/packages/')
theline <- grep("Currently, the CRAN package repository features [0-9]+ available packages.", thepage)
thesplit <- strsplit(thepage[theline], split = " ")[[1]]
thepos <- match("available", thesplit) - 1
n <- thesplit[thepos]
```

> - There are  `r n` available packages on CRAN ([CRAN: Contributed Packages, `r Sys.Date()`](https://cran.r-project.org/web/packages/))
> - R is #5 of all programming languages ([IEEE Spectrum, July 2016](https://www.r-bloggers.com/r-moves-up-to-5th-place-in-ieee-language-rankings/))
> - R is the #1 Google Search for Advanced Analytics software ([Google Trends, March 2014](http://blog.revolutionanalytics.com/2014/03/r-free-popular-powerful-flexible-and-supported.html))
> - R is the highest paid IT skill ([Dice.com survey, January 2014](http://blog.revolutionanalytics.com/2014/02/r-salary-surveys.html))
> - R most-used data science language after SQL ([O'Reilly survey, January 2014](http://blog.revolutionanalytics.com/2014/01/in-data-scientist-survey-r-is-the-most-used-tool-other-than-databases.html))
> - R is used by 70% of data miners ([Rexer survey, October 2013](http://blog.revolutionanalytics.com/2013/10/r-usage-skyrocketing-rexer-poll.html))
> - R growing faster than any other data science language ([KDNuggets survey, August 2013](http://blog.revolutionanalytics.com/2013/09/top-languages-for-data-science.html))


## R: more than just statistics

R was orginally a language __written by statisticians for statisticians__.

Now, it adds more and more features.

## Elegant visualizations tools
- [ggplot2](http://ggplot2.org/)

```{r, echo=TRUE, out.height=330, out.width=530}
library(ggplot2)
dsamp <- diamonds[sample(nrow(diamonds), size = 1000), ]
p <- qplot(carat, price, data = dsamp, colour = clarity)
print(p)
```

## Interactive visualizations tools
- [plotly](https://plot.ly/ggplot2/): interactive plots with the syntax of ggplot2

```{r, eval=FALSE}
library(plotly)
ggplotly(p)
```
<center>
```{r, message=FALSE, out.height=380, out.width=580, echo=FALSE}
library(plotly)
ggplotly(p)
```
</center>

## Use of faster C++ code when needed

> - easy use or C++ code with [Rcpp](https://cran.r-project.org/web/packages/Rcpp/), or other languages
> - advanced (and fast) manipulation of datasets:
    - [data.table](https://cran.r-project.org/web/packages/data.table/): for instance, it enables in-place changes.
    - [dplyr](https://cran.r-project.org/web/packages/dplyr/): fast manipulations of data.frame-like objects.
> - easy parallel computing: [foreach](https://cran.r-project.org/web/packages/foreach/)
> - easy use of on-disk matrices: [bigmemory](https://cran.r-project.org/web/packages/bigmemory/)

You may use JavaScript or C++ code without knowing it.

## Example with foreach and bigmemory

> - Say you have:
    - A SNP big.matrix X stored on-disk in directory _backingfiles_,
    - Infos on the positions of the SNPs (the first 40,000 SNPs are in chromosome 1, then 38,000 are in chromosome 2, etc.),

> - And you have to do some computations which are independent with respect to chromosomes. You want to use __Parallel Computing__!

> - How to do use Parallel Computing on massive genotype matrices?

***

```{r, eval = FALSE, out.height=300}
DO_all <- function(X, infos, ncores) {
  DO_chr <- function(X.desc, lims) {
    X.chr <- sub.big.matrix(X.desc, 
                          firstCol = lims[1], 
                          lastCol = lims[2],
                          backingpath = "backingfiles")
    ## Do something with X.chr
  }
  range.chr <- LimsChr(infos)
  X.desc <- describe(X)
  obj <- foreach(chr = 1:nrow(range.chr), 
                 .packages = "bigmemory")
  expr_fun <- function(chr) {
    DO_chr(X.desc, range.chr[chr, ])
  }
  res <- foreach2(obj, expr_fun, ncores)
}
```

***
```{r, eval = FALSE}
LimsChr <- function(infos) {
  map.rle <- rle(infos$map$chromosome)
  upper <- cumsum(map.rle$length)
  lower <- c(1, upper[-length(upper)] + 1)
  cbind(lower, upper, "chr" = map.rle$values)
}

foreach2 <- function(obj, expr_fun, ncores) {
  if (is.seq <- (ncores == 1)) {
    foreach::registerDoSEQ()
  } else {
    cl <- parallel::makeCluster(ncores)
    doParallel::registerDoParallel(cl)
  }
  res <- eval(parse(
    text = sprintf("foreach::`%%dopar%%`(obj, expr_fun(%s))",
                   obj$argnames)))
  if (!is.seq) parallel::stopCluster(cl)
  return(res)
}
```

## We have RStudio

Live demo

## References and further reading

- [7 Tips For Getting Your Colleagues Hooked on R](http://scl.io/QZxZZl6u#gs.zMhz76Q)
- [Video: What is R?](https://www.youtube.com/embed/TR2bHSJ_eck)
- [How Companies Use R to Compete in a Data-Driven World](http://data-informed.com/companies-use-r-compete-data-driven-world/)
- [How the growth of R helps data-driven organizations succeed](http://www.slideshare.net/RevolutionAnalytics/how-the-growth-of-r-helps-datadriven-organizations-succeed)
- [Seven quick facts about R](http://blog.revolutionanalytics.com/2014/04/seven-quick-facts-about-r.html)